### PR TITLE
Add GRPC Reflection Service to Enhance Dynamic Discovery

### PR DESCRIPTION
**Pull request by Lukaasko**
### Overview 
This pull request introduces a new GRPC reflection service to the plgd-dev/hub project, enhancing the system's ability to dynamically discover GRPC services and methods. This is particularly useful for developers and tools that need to interact with GRPC services without prior knowledge of their interfaces.

### Key Features:

**Service Implementation:**
   - A dedicated GRPC reflection service is implemented, allowing clients to query available GRPC services and their methods dynamically.
   - The service is built using a multi-stage Dockerfile to optimize the build process and final image size.

**Configuration and Deployment:**
   - A new configuration file (`grpc-reflection.yaml`) is introduced, detailing service-specific settings such as logging levels and GRPC addresses.
   - Kubernetes deployment configurations are added, including Helm chart updates to manage the service's lifecycle within a Kubernetes cluster.

**Integration with Existing Infrastructure:**
   - The service is integrated into the existing build and deployment pipelines, ensuring seamless operation alongside other services.
   - Environment variables and configuration options are provided to customize the service's behavior and integration points.

This addition significantly enhances the plgd-dev/hub's capabilities, providing a robust solution for GRPC service discovery and interaction.

### CODE REVIEW

File: charts/plgd-hub/templates/grpc-reflection/_helpers.tpl [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+{{- end }}
+{{- end }}
+{{- end }}
+
+{{- define  "plgd-hub.grpcreflection.image" -}}
+    {{- $registryName := .Values.grpcreflection.image.registry | default "" -}}
+    {{- $repositoryName := .Values.grpcreflection.image.repository -}}
+    {{- $tag := .Values.grpcreflection.image.tag | default .Chart.AppVersion | toString -}}
+    {{- printf "%s%s:%s" $registryName $repositoryName  $tag -}}
+{{- end -}}
```
Comment (jkralik) [inline charts/plgd-hub/templates/grpc-reflection/_helpers.tpl]: pls remove - not used

File: charts/plgd-hub/templates/grpc-reflection/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            - name: {{ .Values.grpcreflection.config.volume }}
+              mountPath: {{ .Values.grpcreflection.config.mountPath }}
+            {{- if ( include "plgd-hub.grpcreflection.createServiceCertByCm" . ) }}
+            - name: service-crt
+              mountPath: {{ $cert }}
+            {{- end }}
+            {{- if .Values.global.authorizationCAPool }}
+            - name: {{ .Values.extraAuthorizationCAPool.name }}
+              mountPath: {{ .Values.extraAuthorizationCAPool.mountPath }}
+            {{- end }}
```
Comment (jkralik) [inline charts/plgd-hub/templates/grpc-reflection/deployment.yaml]: pls replace them with `{{- include "plgd-hub.extraCAPoolMount" (list . .Values.extraCAPool.internal) | nindent 12 }}`

File: charts/plgd-hub/templates/grpc-reflection/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        {{- if ( include "plgd-hub.grpcreflection.createServiceCertByCm" . ) }}
+        - name: service-crt
+          secret:
+            secretName: {{ include "plgd-hub.grpcreflection.serviceCertName" . }}
+        {{- end }}
+        {{- if .Values.global.authorizationCAPool }}
+        - name: {{ .Values.extraAuthorizationCAPool.name }}
+          secret:
+            secretName: {{ .Values.extraAuthorizationCAPool.name }}
+        {{- end }}
```
Comment (jkralik) [inline charts/plgd-hub/templates/grpc-reflection/deployment.yaml]: replace with  `{{- include "plgd-hub.extraCAPoolVolume" (list . .Values.extraCAPool.internal) | nindent 8 }}`

File: charts/plgd-hub/templates/grpc-reflection/config.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+          maxConnectionIdle: {{ .apis.grpc.keepAlive.maxConnectionIdle }}
+          # 0s - means infinity
+          maxConnectionAge: {{ .apis.grpc.keepAlive.maxConnectionAge }}
+          # 0s - means infinity
+          maxConnectionAgeGrace: {{ .apis.grpc.keepAlive.maxConnectionAgeGrace }}
+          time: {{ .apis.grpc.keepAlive.time }}
+          timeout: {{ .apis.grpc.keepAlive.timeout }}
+        tls:
+          {{- $tls := .apis.grpc.tls }}
+          {{- include "plgd-hub.certificateConfig" (list $ $tls $grpcReflectionCertPath) | indent 8 }}
```
Comment (jkralik) [inline charts/plgd-hub/templates/grpc-reflection/config.yaml]: replace with `{{- include "plgd-hub.internalCertificateConfig" (list $ $tls $grpcReflectionCertPath) | indent 8 }}`

File: charts/plgd-hub/values.yaml [hunk 5]
```diff
@@ -2016,6 +2016,161 @@ grpcgateway:
...
+      tls:
+        caPool:
+        keyFile:
+        certFile:
+        clientCertificateRequired: false
+      reflectedServices:
+        - grpcgateway.pb.GrpcGateway
+        - certificateauthority.pb.CertificateAuthority
+        - snippetservice.pb.SnippetService
+
```
Comment (jkralik) [inline charts/plgd-hub/values.yaml]: In few days 1-2 there will be also added ` m2moauthserver.pb.M2MOAuthService`

File: charts/plgd-hub/values.yaml [hunk 6]
```diff
@@ -2016,6 +2016,161 @@ grpcgateway:
...
+  # -- Affinity definition
+  affinity: {}
+  # -- Toleration definition
+  tolerations: {}
+  # -- Optional extra volumes
+  extraVolumes: {}
+  # -- Optional extra volume mounts
+  extraVolumeMounts: {}
+  # -- External domain for grpc-reflection. Default: api.{{ global.domain }}
+  domain:
```
Comment (jkralik) [inline charts/plgd-hub/values.yaml]: Reflection services need to be registered for each domain according to domains of services (each can run on different domain or all/some of them can run on same domain):
- grpcgateway.pb.GrpcGateway
- certificateauthority.pb.CertificateAuthority
- snippetservice.pb.SnippetService
- m2moauthserver.pb.M2MOAuthService

so for each domain we need to create deployment and pod with certain reflected services.

File: tools/grpc-reflection/Makefile [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+build-servicecontainer:
+	$(call build-docker-image,service)
+
+build: build-servicecontainer
+
+push: build-servicecontainer
+	docker push plgd/$(SERVICE_NAME):$(VERSION_TAG)
+	docker push plgd/$(SERVICE_NAME):$(LATEST_TAG)
+
+proto/generate:
```
Comment (Danielius1922) [inline tools/grpc-reflection/Makefile:37]: remove if there is no protobuf to generate (also on line 40)

File: charts/plgd-hub/templates/grpc-reflection/_helpers.tpl (hunk 1 above)
Comment (Lukaasko) [reply charts/plgd-hub/templates/grpc-reflection/_helpers.tpl]: removed

File: charts/plgd-hub/templates/grpc-reflection/deployment.yaml (hunk 2 above)
Comment (Lukaasko) [reply charts/plgd-hub/templates/grpc-reflection/deployment.yaml]: replaced

File: charts/plgd-hub/templates/grpc-reflection/deployment.yaml (hunk 3 above)
Comment (Lukaasko) [reply charts/plgd-hub/templates/grpc-reflection/deployment.yaml]: replaced

File: charts/plgd-hub/templates/grpc-reflection/config.yaml (hunk 4 above)
Comment (Lukaasko) [reply charts/plgd-hub/templates/grpc-reflection/config.yaml]: replaced

File: tools/grpc-reflection/config.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    timeEncoder: rfc3339nano
+
+apis:
+  grpc:
+    address: "0.0.0.0:9100"
+    reflectedServices:
+      - grpcgateway.pb.GrpcGateway
+      - certificateauthority.pb.CertificateAuthority
+      - snippetservice.pb.SnippetService
+      - m2moauthserver.pb.M2MOAuthService
```
Comment (Danielius1922) [inline tools/grpc-reflection/config.yaml:17]: will dps be reflected?

File: charts/plgd-hub/values.yaml [hunk 9]
```diff
@@ -2024,6 +2024,154 @@ grpcgateway:
...
+    annotations:
+      nginx.org/grpc-services: '{{ include "plgd-hub.grpcreflection.fullname" . }}'
+      nginx.ingress.kubernetes.io/backend-protocol: "GRPCS"
+      ingress.kubernetes.io/force-ssl-redirect: "true"
+      nginx.ingress.kubernetes.io/enable-cors: "true"
+      cert-manager.io/private-key-rotation-policy: always
+    # -- Custom map of Ingress annotation
+    customAnnotations: {}
+    paths:
+      - /grpc.reflection.v1alpha.ServerReflection
```
Comment (Danielius1922) [inline charts/plgd-hub/values.yaml:2122]: do we need the alpha path?

File: charts/plgd-hub/templates/grpc-reflection/_helpers.tpl [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+{{- define "plgd-hub.grpcreflection.domain" -}}
+  {{- if .Values.grpcreflection.domain }}
+    {{- printf "%s" .Values.grpcreflection.domain }}
+  {{- else }}
+    {{- printf "api.%s" .Values.global.domain }}
+  {{- end }}
+{{- end }}
+
+
+{{- define "plgd-hub.grpcreflection.internalDns" -}}
```
Comment (Danielius1922) [inline charts/plgd-hub/templates/grpc-reflection/_helpers.tpl]: doesn't seem to be included anywhere when I search by grep (same for plgd-hub.grpcgateway.internalDns)

Comment (Danielius1922) [reply charts/plgd-hub/templates/grpc-reflection/_helpers.tpl]: maybe it should be used for `spec.dnsNames` in  `charts/plgd-hub/templates/grpc-reflection/service-crt.yaml` ?

File: snippet-service/service/service.go [hunk 11]
```diff
@@ -77,6 +77,9 @@ func newHttpService(ctx context.Context, config HTTPConfig, validatorConfig vali
 }
 
 func newGrpcService(ctx context.Context, config grpcService.Config, ss *grpcService.SnippetServiceServer, fileWatcher *fsnotify.Watcher, logger log.Logger, tracerProvider trace.TracerProvider) (*grpcService.Service, func(), error) {
+	if err := config.Authorization.Validate(); err != nil {
```
Comment (Danielius1922) [inline snippet-service/service/service.go]: This seems a bit weird, no other config has the `Validate` method called inside the service.go why is it needed here?

File: charts/plgd-hub/templates/grpc-reflection/_helpers.tpl (hunk 10 above)
Comment (jkralik) [reply charts/plgd-hub/templates/grpc-reflection/_helpers.tpl]: removed

File: snippet-service/service/service.go (hunk 11 above)
Comment (jkralik) [reply snippet-service/service/service.go]: fixed

**Comment by jkralik** [comment]
@coderabbitai review

**Comment by jkralik** [comment]
@coderabbitai create pull-request description for non-plgd developers

**Comment by jkralik** [comment]
@coderabbitai create pull-request description focused to grpc-reflection for non-plgd developers

**Review by Danielius1922 — APPROVED** [review]

**Pull request merged**